proposals/20220823-huggingface_model_pusher.md

```diff
@@ -0,0 +1,74 @@
+#### SIG TFX-Addons
+# Project Proposal
+
+---
+
+**Your name:** Chansung Park
+
+**Your email:** deep.diver.csp@gmail.com
+
+**Your company/organization:** Individual ([ML GDE](https://developers.google.com/community/experts/directory/profile/profile-chansung-park))
+
+**Project name:** HuggingFace Model Pusher
+
+## Project Description
+HuggingFace Model Pusher(`HFModelPusher`) pushes blessed model to the [HuggingFace Model Hub](https://huggingface.co/models).
+
+## Project Category
+Component
+
+## Project Use-Case(s)
+The HuggingFace Model Hub lets us have [Git-LFS](https://git-lfs.github.com) enabled repositories in public and private modes. Supported models hosted on the HuggingFace Model Hub can be directly loaded/used with APIs provided by [transformers](https://huggingface.co/docs/transformers/index) package. However, it is not limited. We can host arbitrary types of models too. 
+
+HuggingFace Model Hub is easy to manage model versions, especially for those familiar with Git.
+
+## Project Implementation
+HFModelPusher is a class-based TFX component, and it inherits from TFX standard `Pusher` component.
+
+It takes the following inputs:
```

Comment:
It might be nice to include a README and/or a model_card_metadata [config](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1) as inputs for additional documentation and discoverability.

Reply:
@codesue 

Thank you for the suggestion! me and @sayakpaul had the same thought. 

Specifically, it would be great to upload a model card generated by `Evaluator` and [`model-card-toolkit`](https://github.com/tensorflow/model-card-toolkit). However the `model-card-toolkit` is a on-going project to be ported into TFX Add-on, so I thought maybe upgrade `HFModelPusher` when `model-card-toolkit` is completed.

By the way, your suggestion on the [model_card_metadata config](https://github.com/huggingface/hub-docs/blob/main/modelcard.md?plain=1) sounds good too!  But, it has many many information to fill in, so it would be inappropriate for a TFX component to fill in automatically. Do you have any idea how to make things easier for users to use this?

Reply:
Similar to how mct fills information, it may be good to automatically fill some of this by adding Statistics and such as inputs to the component. See the existing ModeCardGenerator component for reference -> https://github.com/tensorflow/model-card-toolkit/blob/master/model_card_toolkit/tfx/executor.py

Reply:
so, taking outputs from `Evaluator` and `StatisticsGen` optionally.
- if both Artifacts exist, create a `ModelCardToolkit`
- then fill some information from it

Reply:
Yeah, that sounds good! Not all the HF model card fields are available in MCT, and MCT doesn't support custom fields yet, so we won't be able to completely fill in the HF config section automatically, but there's still a lot that can be automatically annotated if we use a custom template.

Reply:
The outputs of `Trainer` doesn’t have metrics information, and evaluation is done with `Evaluator`. 

So, we can include very general information such as display name, tags, which dataset was used, or something like it. We can prepare a template file filled up with default values, then let them overwrite if any inputs are given.

Reply:
I think [`modelcards`](https://github.com/nateraw/modelcards) package is useful for this purpose.

Reply:
@sayakpaul @codesue @casassg 

Or it would be easier to just put HTML contents generated by MCT into the markdown model card in HuggingFace Mode Repo. WDYT?

Reply:
I think we would want to avoid the code for turning an HTML page into a separate markdown file for the Hugging Face Hub README. IMO, we could develop separate utilities for this purpose:

* Have a template for the model card (TFX and HF have their own formats, it seems, but there's overlap in information).
* Have a utility that generates HTML page from the populated template text 
* Have utility that generates a markdown file from the populated template text

Reply:
Right. 

Since this component could be over complicated if we include Model Card generation feature, I thought two possible solutions:
1. Create another custom TFX evaluation component for HuggingFace Model 
2. Or just run `evaluate` inside this component(`HFModelPusher`) over the test dataset and fill evaluation results into the Model Card
3. Or leverage the existing `Evaluator` standard TFX component

Reply:
Model Card Toolkit can also create markdown or any arbitrary file types since you can have custom templates and filenames, so it wouldn't require converting HTML to markdown. Anyway, I'm in favor of your and @sayakpaul's original idea to upgrade `HFModelPusher` to include a model card later in order to prevent increasing scope and to avoid adding dependencies and unknowns.